File: framework-web/src/main/java/com/example/framework/web/interceptor/TraceIdInterceptor.java
package com.example.framework.web.interceptor;

import com.example.framework.web.constants.WebConstants;
import io.micrometer.tracing.TraceContext;
import io.micrometer.tracing.Tracer;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.AllArgsConstructor;
import org.jetbrains.annotations.NotNull;
import org.springframework.web.servlet.HandlerInterceptor;

/**
 * @author MaGuangZu
 * @since 2024-05-19
 */
@AllArgsConstructor
public class TraceIdInterceptor implements HandlerInterceptor {

	private final Tracer tracer;

	@Override
	public boolean preHandle(@NotNull HttpServletRequest request, @NotNull HttpServletResponse response, @NotNull Object handler) throws Exception {
		TraceContext context = tracer.currentTraceContext().context();
		if (context != null && context.traceId() != null) {
			response.setHeader(WebConstants.TRACE_ID_KEY, context.traceId());
		}
		return HandlerInterceptor.super.preHandle(request, response, handler);
	}

}

File: framework-web/src/main/kotlin/com/example/framework/web/interceptor/TraceIdInterceptor.kt
package com.example.framework.web.interceptor

import com.example.framework.web.constants.WebConstants.TRACE_ID_KEY
import io.micrometer.tracing.Tracer
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import lombok.extern.slf4j.Slf4j
import org.springframework.web.servlet.HandlerInterceptor

/**
 * trace id interceptor
 *
 * @author MaGuangZu
 * @since 2021-12-01
 */
@Slf4j
class TraceIdInterceptor(private val tracer: Tracer) : HandlerInterceptor {

	@Throws(Exception::class)
    override fun preHandle(request: HttpServletRequest, response: HttpServletResponse, handler: Any): Boolean {
		response.setHeader(TRACE_ID_KEY, tracer.currentTraceContext().context()?.traceId())
        return super.preHandle(request, response, handler)
    }

}
